server/mqtt.go

```diff
@@ -3140,44 +3298,231 @@ func mqttPubTrace(pp *mqttPublish) string {
-
-	bb := bytes.Buffer{}
-	bb.WriteString(hdrLine)
-	bb.Write(mqttNatsHeaderB)
-	bb.WriteByte(':')
-	bb.WriteByte('0' + mqttGetQoS(pp.flags))
-	bb.WriteString(_CRLF_)
-	bb.WriteString(_CRLF_)
-	c.pa.hdr = bb.Len()
-	c.pa.hdb = []byte(strconv.FormatInt(int64(c.pa.hdr), 10))
-	bb.Write(pp.msg)
-	c.pa.size = bb.Len()
-	c.pa.szb = []byte(strconv.FormatInt(int64(c.pa.size), 10))
-	msgToSend := bb.Bytes()
+func (s *Server) mqttProcessPub(c *client, pp *mqttPublish, trace bool) error {
+	qos := mqttGetQoS(pp.flags)
 
-	var err error
-	// Unless we have a publish permission error, if the message is QoS1, then we
-	// need to store the message (and deliver it to JS durable consumers).
-	if _, permIssue := c.processInboundClientMsg(msgToSend); !permIssue && mqttGetQoS(pp.flags) > 0 {
-		// We need to call flushClients now since this we may have called c.addToPCD
-		// with destination clients (possibly a route). Without calling flushClients
-		// the following call may then be stuck waiting for a reply that may never
-		// come because the destination is not flushed (due to c.out.fsp > 0,
-		// see addToPCD and writeLoop for details).
-		c.flushClients(0)
-		// Store this QoS1 message.
-		_, err = c.mqtt.sess.jsa.storeMsg(mqttStreamSubjectPrefix+string(c.pa.subject), c.pa.hdr, msgToSend)
-	}
-	c.pa.subject, c.pa.mapped, c.pa.hdr, c.pa.size, c.pa.szb, c.pa.reply = nil, nil, -1, 0, nil, nil
+	switch qos {
+	case 0, 1:
+		err := s.mqttInitiateMsgDelivery(c, pp)
+		if qos == 1 {
+			// Send a PUBACK back to the client.
+			c.mqttEnqueuePubResponse(mqttPacketPubAck, pp.pi, trace)
+		}
+		return err
+
+	case 2:
+		err := s.mqttStoreQOS2Msg(c, pp)
+		// Send a PUBREC back to the client.
+		c.mqttEnqueuePubResponse(mqttPacketPubRec, pp.pi, trace)
+		return err
+
+	default:
+		return fmt.Errorf("unreachable: invalid QoS in mqttProcessPub: %v", qos)
+	}
+}
+
+func (s *Server) mqttInitiateMsgDelivery(c *client, pp *mqttPublish) error {
+	natsMsg, headerLen := mqttNewDeliverableMessage(pp, false)
+
+	// Broadcast the message just as we do for QoS0, but this message is
+	// recognized as QoS1 (stored in the NATS header), so it will be ignored
+	// by the QoS1+ subscriptions. Those will expect the message to be
+	// delivered via JetStream.
+	permIssue := s.mqttDeliverNATS(c, pp, natsMsg, headerLen)
+	if permIssue {
+		return nil
+	}
+
+	// <>/<> what is addToPCD?
+	// We need to call flushClients now since this we may have called c.addToPCD
+	// with destination clients (possibly a route). Without calling flushClients
+	// the following call may then be stuck waiting for a reply that may never
+	// come because the destination is not flushed (due to c.out.fsp > 0,
+	// see addToPCD and writeLoop for details).
+	c.flushClients(0)
+
+	// Store the message in JetStream as "$MQTT.msgs.<delivery-subject>"
+	_, err := c.mqtt.sess.jsa.storeMsg(mqttStreamSubjectPrefix+string(pp.subject), headerLen, natsMsg)
```

Comment:
I checking to see if there is not cases where we would come with c.pa.subject being something and that being changed in processInboundClientMsg. If that is the case, we would store here with the wrong subject (original code referenced c.pa.subject).
We do possibly change c.pa.subject in case of gateways, so again, not sure if the refactoring you did is safe...

Reply:
So what I would suggest is that since `mqttDeliverNATS` (the function that invokes process inbound but has a defer that clears all `c.pa.xxx` fields (therefore `c.pa.subject`) is only invoked - so far - in `mqttInitiateMsgDelivery`, you could do the defer that clears c.pa fields in that function (before calling `mqttDeliverNATS`) and use `c.pa.subject` here instead of `pp.subject`, or do not clear `c.pa.subject` has part of the defer inside `mqttDeliverNATS`, but then it means still have a defer here to make sure you clear it since there are several returns here.

(Also, since this is always invoked from the client's readLoop, and we set c.pa.subject=pp.subject at the beginning of the processing, we *could* simply not do the reset in the defer. But I wanted to do that because pp.subject may sometime reference the read buffer, so I wanted to make sure that this would not cause the compiler to have to make a copy because the reference would still be held after a read loop iteration).

Reply:
@kozlovic Restored the exact(?) logic from before, including an explicit `flushClient` in `mqttHandleWill`.